src/arraymancer/tensor/selectors.nim

```diff
@@ -255,17 +247,48 @@ proc masked_fill*[T](t: var Tensor[T], mask: openArray, value: Tensor[T]) =
   ##
   ## Example:
   ##
-  ##   t.masked_fill([true, false, true, true], [3, 4, -1].toTensor)
+  ##   t.masked_fill(t > 0, [3, 4, -1].toTensor)
   ##
-  ## In this version of this procedure the boolean mask, which must have the
-  ## same size as the input tensor ``t``, is an openArray of bools, i.e.:
-  ##   - an array or sequence of bools
-  ##   - an array of arrays of bools,
-  ##   - ...
+  ## In this version of this procedure the boolean mask is a ``Tensor[bool]``
+  ## with the same size as the input tensor ``t``.
+  masked_fill_impl(t, mask, value)
+
+proc masked_fill*[T](t: var Tensor[T], mask: Tensor[bool], value: openArray[T]) =
+  ## Version of `masked_fill` that takes an openArray as the value
+  ##
+  ## For each element ``t[index]`` of the input tensor ``t`` with index ``index``,
+  ## check if ``mask[index]`` is true. If so fill it with the _next_
+  ## element from the ``value`` openArray. Otherwise leave it untouched.
+  ##
+  ## Note that this does _not_ fill ``t[index]`` with ``value[index]``, but
+  ## with the n-th element of ``value`` where n is the number of true elements
+  ## in the mask before and including the index-th mask element.
+  ## Because of this, the value openArray must have at least as many elements as
+  ## the number of true elements in the mask. If that is not the case an
+  ## IndexDefect exception will be raised at runtime. The ``value`` tensor
+  ## can have even more values which will simply be ignored.
+  ##
+  ## Example:
+  ##
+  ##   t.masked_fill(t > 0, [3, 4, -1])
+  ##
+  ## In this version of this procedure the boolean mask is a ``Tensor[bool]``
+  ## with the same size as the input tensor ``t``.
+  masked_fill_impl(t, mask, value)
 
+proc masked_fill*[T](t: var Tensor[T], mask: openArray, value: Tensor[T]) =
```

Comment:
Does explicit `openArray[bool]` break?

Reply:
Yes, openArray[bool] does not work because the mask can be, for example, a seq[seq[bool]]. Note the aim not changing this, it was already like that in the orirginal code.

Reply:
Yup, but I forgot to ask I think. Good to know and that makes sense! :)